exec/java-exec/src/main/java/org/apache/drill/exec/rpc/user/UserServer.java

```diff
@@ -246,28 +154,80 @@ protected void handle(UserClientConnectionImpl connection, int rpcType, ByteBuf
   public class UserClientConnectionImpl extends RemoteConnection implements UserClientConnection {
 
     private UserSession session;
+    private SaslServer saslServer;
+    private RequestHandler<UserClientConnectionImpl> currentHandler;
+    private UserToBitHandshake inbound;
 
     public UserClientConnectionImpl(SocketChannel channel) {
       super(channel, "user client");
+      currentHandler = authFactory == null ? handler : new UserServerAuthenticationHandler(handler);
     }
 
     void disableReadTimeout() {
       getChannel().pipeline().remove(BasicServer.TIMEOUT_HANDLER);
     }
 
-    void setUser(final UserToBitHandshake inbound) throws IOException {
+    void setHandshake(final UserToBitHandshake inbound) throws IOException {
+      this.inbound = inbound;
+    }
+
+    void initSaslServer(final String mechanismName, final Map<String, ?> properties)
+        throws IllegalStateException, SaslException {
+      if (saslServer != null) {
+        throw new IllegalStateException("SASL server already initialized.");
+      }
+      this.saslServer = authFactory.getMechanism(mechanismName)
```

Comment:
It should be the responsibility of class AuthenticationMechanismFactory to decide what mechanism to provide. Right? So AuthenticationMechanismFactory may also need 'properties' to decide what mechanism to ask for. E.g. Some company may want connections from out of VPN or in VPN to use different mechanisms.